derive_builder_core/src/try_from.rs

```diff
@@ -0,0 +1,174 @@
+    pub builder_ty: &'a syn::Ident,
+    /// Name of the build fn.
+    pub fn_ident: syn::Ident,
+    /// How the build method takes and returns `self` (e.g. mutably).
+    pub pattern: BuilderPattern,
+    /// Type of the target field.
+    ///
+    /// The corresonding builder field will be `Option<field_type>`.
+    pub target_ty: &'a syn::Ident,
+    /// Type parameters and lifetimes attached to this builder struct.
+    pub generics: &'a syn::Generics,
+    
+    /// Bindings to libstd or libcore.
+    pub bindings: Bindings,
+}
+
+impl<'a> ToTokens for TryFromImpl<'a> {
+    fn to_tokens(&self, tokens: &mut Tokens) {
+        if !self.enabled {
+            return;
+        }
+        
+        // For the generic declarations after impl, we need a
+        // lifetime so that we can declare TryFrom<&'whatever FooBuilder>.
+        // This should NOT appear in the where clause or target type declaration,
+        // which is why we have to make the clone.
+        let mut i_generics = self.generics.clone();
+        i_generics.lifetimes.push(syn::LifetimeDef::new("'unused"));
+        let (impl_generics, _, _) = i_generics.split_for_impl();
+        
+        let builder_ty = &self.builder_ty;
+        let target_ty = &self.target_ty;
+        let (owned_generics, ty_generics, where_clause) = self.generics.split_for_impl();
+        let fn_ident = &self.fn_ident;
+        let try_from = self.bindings.try_from_trait();
+        let result_ty = self.bindings.result_ty();
+        let string_ty = self.bindings.string_ty();
+        
+        let immutable_borrow = quote!(
+            impl #impl_generics #try_from<&'unused #builder_ty #ty_generics> for #target_ty #ty_generics 
+                #where_clause {
+                type Error = #string_ty;
+                
+                fn try_from(v: &#builder_ty #ty_generics) -> #result_ty<Self, Self::Error> {
+                    v.#fn_ident()
+                }
+            }
+        );
+        
+        let mutable_borrow = quote!(
+            impl #impl_generics #try_from<&'unused mut #builder_ty #ty_generics> for #target_ty #ty_generics 
+                #where_clause {
+                type Error = #string_ty;
+                
+                fn try_from(v: &mut #builder_ty #ty_generics) -> #result_ty<Self, Self::Error> {
+                    v.#fn_ident()
+                }
+            }
+        );
+        
+        let owned = quote!(
+            impl #owned_generics #try_from<#builder_ty #ty_generics> for #target_ty #ty_generics 
+                #where_clause {
+                type Error = #string_ty;
+                
+                fn try_from(v: #builder_ty #ty_generics) -> #result_ty<Self, Self::Error> {
+                    v.#fn_ident()
+                }
+            }
+        );
+        
+        tokens.append(quote!(
+            #immutable_borrow
+            
+            #mutable_borrow
+            
+            #owned            
```

Comment:
I'd like to start with only one impl depending on the builder pattern

# Variant 1: Mimic `build()` method

Pseudo code for demonstration:

```rust
Owned => /* ... impl TryFrom<FooBuilder> for Foo ... */,
Mutable |
Immutable => /* ... impl TryFrom<&FooBuilder> for Foo ... */,
```

So despite having three variants of `BuilderPattern` there would only be two variants of `TryFrom` implementations - in analogy to the `BuildMethod`.

# Variant 2: Always take `Self`

The logic behind changing the signature of the build method was to enable method chaining `FooBuilder::default().foo(1).build()`. But this would most likely not be the primary use case of `TryFrom` - although it _could_ be used in this position.

Alternatively we could talk about _always_ implementing `TryFrom<FooBuilder> for Foo` and _never_ implementing `TryFrom<&FooBuilder> for Foo`.

# Discussion

We should probably discuss the primary use cases for having `TryFrom` implementations. What signature would they require? Would autoderef or other coercions help to pick the right `TryFrom` implementation?